File: src/main/java/org/apache/fineract/client/services/ClientsAddressApi.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.fineract.client.services;

import org.apache.fineract.client.CollectionFormats.*;

import io.reactivex.Observable;
import io.reactivex.Completable;
import retrofit2.Call;
import retrofit2.http.*;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.MultipartBody;

import org.apache.fineract.client.models.GetClientClientIdAddressesResponse;
import org.apache.fineract.client.models.PostClientClientIdAddressesRequest;
import org.apache.fineract.client.models.PostClientClientIdAddressesResponse;
import org.apache.fineract.client.models.PutClientClientIdAddressesRequest;
import org.apache.fineract.client.models.PutClientClientIdAddressesResponse;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public interface ClientsAddressApi {
  /**
   * Create an address for a Client
   * Mandatory Fields :  type and clientId
   * @param clientid clientId (required)
   * @param postClientClientIdAddressesRequest  (required)
   * @param type type (optional)
   * @return Call&lt;PostClientClientIdAddressesResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @POST("client/{clientid}/addresses")
  Call<PostClientClientIdAddressesResponse> addClientAddress(
    @retrofit2.http.Path("clientid") Long clientid, @retrofit2.http.Body PostClientClientIdAddressesRequest postClientClientIdAddressesRequest, @retrofit2.http.Query("type") Long type
  );

  /**
   * List all addresses for a Client
   * Example Requests:  client/1/addresses   clients/1/addresses?status&#x3D;false,true&amp;&amp;type&#x3D;1,2,3
   * @param clientid clientId (required)
   * @param status status (optional)
   * @param type type (optional)
   * @return Call&lt;List&lt;GetClientClientIdAddressesResponse&gt;&gt;
   */
  @GET("client/{clientid}/addresses")
  Call<List<GetClientClientIdAddressesResponse>> getAddresses1(
    @retrofit2.http.Path("clientid") Long clientid, @retrofit2.http.Query("status") String status, @retrofit2.http.Query("type") Long type
  );

  /**
   * 
   * 
   * @return Call&lt;String&gt;
   */
  @GET("client/addresses/template")
  Call<String> getAddressesTemplate();
    

  /**
   * Update an address for a Client
   * All the address fields can be updated by using update client address API  Mandatory Fields type and addressId
   * @param clientid clientId (required)
   * @param putClientClientIdAddressesRequest  (required)
   * @return Call&lt;PutClientClientIdAddressesResponse&gt;
   */
  @Headers({
    "Content-Type:application/json"
  })
  @PUT("client/{clientid}/addresses")
  Call<PutClientClientIdAddressesResponse> updateClientAddress(
    @retrofit2.http.Path("clientid") Long clientid, @retrofit2.http.Body PutClientClientIdAddressesRequest putClientClientIdAddressesRequest
  );

}

File: src/main/kotlin/org/openapitools/client/apis/ClientsAddressApi.kt
package org.openapitools.client.apis

import org.openapitools.client.models.GetClientClientIdAddressesResponse
import org.openapitools.client.models.PostClientClientIdAddressesRequest
import org.openapitools.client.models.PostClientClientIdAddressesResponse
import org.openapitools.client.models.PutClientClientIdAddressesRequest
import org.openapitools.client.models.PutClientClientIdAddressesResponse
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

interface ClientsAddressApi {
    /**
     * Create an address for a Client
     * Mandatory Fields :  type and clientId
     * Responses:
     *  - 200: OK
     *
     * @param clientid clientId
     * @param postClientClientIdAddressesRequest
     * @param type type (optional)
     * @return [PostClientClientIdAddressesResponse]
     */
    @POST("v1/client/{clientid}/addresses")
    suspend fun addClientAddress(
        @Path("clientid") clientid: Long,
        @Body postClientClientIdAddressesRequest: PostClientClientIdAddressesRequest,
        @Query("type") type: Long? = null
    ): PostClientClientIdAddressesResponse

    /**
     * List all addresses for a Client
     * Example Requests:  client/1/addresses   clients/1/addresses?status&#x3D;false,true&amp;&amp;type&#x3D;1,2,3
     * Responses:
     *  - 200: OK
     *
     * @param clientid clientId
     * @param status status (optional)
     * @param type type (optional)
     * @return [kotlin.collections.List<GetClientClientIdAddressesResponse>]
     */
    @GET("v1/client/{clientid}/addresses")
    suspend fun getAddresses1(
        @Path("clientid") clientid: Long,
        @Query("status") status: String? = null,
        @Query("type") type: Long? = null
    ): List<GetClientClientIdAddressesResponse>

    /**
     *
     *
     * Responses:
     *  - 0: default response
     *
     * @return [kotlin.String]
     */
    @GET("v1/client/addresses/template")
    suspend fun getAddressesTemplate(): String

    /**
     * Update an address for a Client
     * All the address fields can be updated by using update client address API  Mandatory Fields type and addressId
     * Responses:
     *  - 200: OK
     *
     * @param clientid clientId
     * @param putClientClientIdAddressesRequest
     * @return [PutClientClientIdAddressesResponse]
     */
    @PUT("v1/client/{clientid}/addresses")
    suspend fun updateClientAddress(
        @Path("clientid") clientid: Long,
        @Body putClientClientIdAddressesRequest: PutClientClientIdAddressesRequest
    ): PutClientClientIdAddressesResponse

}
